vm-virtio/src/console.rs

```diff
@@ -0,0 +1,353 @@
+            DeviceError::FailedSignalingUsedQueue(e)
+        })
+    }
+
+    fn run(&mut self) -> result::Result<(), DeviceError> {
+        // Create the epoll file descriptor
+        let epoll_fd = epoll::create(true).map_err(DeviceError::EpollCreateFd)?;
+
+        // Add events
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.input_queue_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(INPUT_QUEUE_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.output_queue_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(OUTPUT_QUEUE_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.kill_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(KILL_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+
+        const EPOLL_EVENTS_LEN: usize = 100;
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); EPOLL_EVENTS_LEN];
+
+        'epoll: loop {
+            let num_events =
+                epoll::wait(epoll_fd, -1, &mut events[..]).map_err(DeviceError::EpollWait)?;
+
+            for event in events.iter().take(num_events) {
+                let ev_type = event.data as u16;
+
+                match ev_type {
+                    INPUT_QUEUE_EVENT => {
+                        if let Err(e) = self.input_queue_evt.read() {
+                            error!("Failed to get queue event: {:?}", e);
+                            break 'epoll;
+                        } else if self.process_input_queue() {
+                            if let Err(e) = self.signal_used_queue() {
+                                error!("Failed to signal used queue: {:?}", e);
+                                break 'epoll;
+                            }
+                        }
+                    }
+                    OUTPUT_QUEUE_EVENT => {
+                        if let Err(e) = self.output_queue_evt.read() {
+                            error!("Failed to get queue event: {:?}", e);
+                            break 'epoll;
+                        } else {
+                            self.process_output_queue();
+                        }
+                    }
+                    KILL_EVENT => {
+                        debug!("KILL_EVENT received, stopping epoll loop");
+                        break 'epoll;
+                    }
+                    _ => {
+                        error!("Unknown event for virtio-block");
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Virtio device for exposing console to the guest OS through virtio.
+pub struct Console {
```

Comment:
This structure should be decoupled into two distinct ones. The goal being to decouple virtio specific things from input specific things. This is a bit long to explain but this device could be directly owned by a `VirtioPciDevice`, which would prevent the VMM (`vm.rs`) to own it too. And this would need some changes at the VirtioPciDevice level to accept an `Arc<Mutex<VirtioDevice>>`. This would increase the complexity and this would have to lock everytime some data needs to be modified.
Instead, by using a separate `ConsoleInput` structure, we can avoid this kind of issue, by letting this structure having access to the `in_buffer` and `input_evt` only.

Reply:
Done by @sboeuf.